From eaa37594375e3d6bbd3365ce204edf0f64a10214 Mon Sep 17 00:00:00 2001
From: John Noller <jnoller@anaconda.com>
Date: Mon, 21 Jul 2025 17:34:23 -0400
Subject: [PATCH] hwcap_sve_check

Co-authored-by: Charles Bousseau <cbousseau@anaconda.com>

In https://github.com/ggerganov/llama.cpp/pull/9331 upstream added sve support detection on aarch64.
However, the checks for HWCAP_SVE and HWCAP_ASIMDDP are not available on older kernels / libc.
This patch adds a fallback for systems that lack the HWCAP_SVE check.
This patch also fixes the HWCAP_ASIMDDP check logic (&& -> &)

After updating to glibc 2.28 and kernel 4.18.0, HWCAP2_SVE2 and HWCAP2_SME are the ones not defined 
in the kernel headers.
---
 ggml/src/ggml-cpu/arch/arm/cpu-feats.cpp | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/ggml/src/ggml-cpu/arch/arm/cpu-feats.cpp b/ggml/src/ggml-cpu/arch/arm/cpu-feats.cpp
index 67369147c..956952ad4 100644
--- a/ggml/src/ggml-cpu/arch/arm/cpu-feats.cpp
+++ b/ggml/src/ggml-cpu/arch/arm/cpu-feats.cpp
@@ -33,9 +33,22 @@ struct aarch64_features {
         has_dotprod = !!(hwcap & HWCAP_ASIMDDP);
         has_fp16_va = !!(hwcap & HWCAP_FPHP);
         has_sve     = !!(hwcap & HWCAP_SVE);
-        has_sve2    = !!(hwcap2 & HWCAP2_SVE2);
         has_i8mm    = !!(hwcap2 & HWCAP2_I8MM);
-        has_sme     = !!(hwcap2 & HWCAP2_SME);
+
+    // not defined in kernel headers 4.18.0
+    #if defined(HWCAP2_SVE2)
+        has_sve2 = !!(hwcap2 & HWCAP2_SVE2);
+    #else
+        has_sve2 = false;
+    #endif
+
+    // not defined in kernel headers 4.18.0
+    #if defined(HWCAP2_SME)
+        has_sme = !!(hwcap2 & HWCAP2_SME);
+    #else
+        has_sme = false;
+    #endif
+
 #elif defined(__APPLE__)
         int oldp = 0;
         size_t size = sizeof(oldp);
-- 
2.39.5 (Apple Git-154)

